docs/docs/routing.md

```diff
@@ -0,0 +1,261 @@
+
+Secondly, navigation events are only emitted when they're caused by external stimuli such as the *Back* or *Forward* buttons. When the application itself updates the URL, nothing will happen. This means that the application will need to manually pass the new URL from the link that caused it to the top-level component that is handling routing.
+
+Typically, these issues are solved by creating a separate object which manages access to the HTML5 History API. This object will do two things:
+
+- It will normalise events between browsers
+- It will emit navigation events caused by both the application and external stimuli
+
+While it is certainly possible to create your own object to manage access to the History API, in this guide we'll use the popular [history](https://github.com/mjackson/history) package.
+
+```bash
+npm install --save history
+```
+
+As the History API is global, our object will also need to be accessible everywhere. For example, we can create a separate file and export the object from it, so that any other modules can later import it.
+
+```js
+import createBrowserHistory from 'history/createBrowserHistory';
+
+const browserHistory = createBrowserHistory();
+
+export default browserHistory;
+```
+
+This guide will use the following methods and properties from the `browserHistory` object:
+
+-   `browserHistory.location`
+
+    An object that contains the browser's current location. Its shape is similar to the object available at `window.location`:
+
+    ```js
+    {
+      pathname: '/friends',
+      query: '?page=1',
+      hash: '#top'
+    }
+    ```
+
+-   `browserHistory.listen(callbackFunction)`
+    
+    Subscribes to history events, returning a function that will cancel the subscription when called.
+
+    While subscribed, the callback function will be called with the latest `location` each time the user navigates. For example:
+
+    ```js
+    const unlisten = browserHistory.listen((location) => {
+      console.log(location.pathname);
+    });
+    
+-   `browserHistory.push(location)`
+
+    Adds a new entry to the browser history using `window.history.pushState()`, then calls any listeners registered with `browserHistory.listen()`
+
+You can view the documentation for the full `browserHistory` object API at the [history](https://github.com/mjackson/history) package's page.
+
+## Links
+
+By default, HTML `<a>` elements will always cause the page to reload when clicked. If you'd like to update the URL using the History API, you'll need to cancel the default behaviour and then implement an alternative.
+
+You could do this by adding an `onClick` handler to each individual `<a>` element, but since most links within an app will share the same logic, people often create a component for them. Instances of this component behave exactly like `<a>` elements, except that instead of refreshing the page, they will update the URL in-place using `browserHistory.push()`:
+
+```js{47}
+class Link extends React.Component {
```

Comment:
To close the whole topic revolving around hurt feelings by `react-router`, I propose simplifying the guide.

As we are building a basic routing solution first, we can remove all the nitty-gritty details of handing `onClick`. Start simple: do a `e.preventDefault()` and call some transition function unconditionally:

> ```js
>  handleClick(event) {
>     event.preventDefault();
>     // Instead of a normal URL, browserHistory.push() expects an object with
>     // `pathname` and `search` properties
>     const pathname = this.props.href.split('?')[0];
>     const search = this.props.href.slice(pathname.length);
>     const location = {pathname, search};
>     
>     // Update the browser URL and notify any listeners added
>     // via browserHistory.listen()
>    browserHistory.push(location);
>  }
> ```

This will be a very basic working router implementation. It's short, it's sweet, it works, it's perfect for a beginner.

After this is done, after other pieces like rendering proper components in `App` are explained, briefly explain other changes you can make to the `Link` component:

> To extend `Link` functionality, you may want to extend `handleClick` to correctly work in a variety of situations:
> - support for `onClick={...}` may be handled by `if(this.props.onClick) ...`
> - check if the user clicked left button, not right button: `const isNotLeftClick = event.button !== 0;`

And conclude with the paragraph I proposed [here](https://github.com/facebook/react/pull/8848#discussion_r105359987)